### PR TITLE
Add ability to set keys on the global blackboard

### PR DESCRIPTION
**Pull request by veerajsk**
- The master plan can specify if its keys are meant for the global blackboard using the inherit_blackboard flag in the .plc file. If true, the keys are populated to the global blackboard & the master plan simply inherits the global blackboard as its own. If false, the behaviour remains unchanged.
- Blackboard keys can now have default values

Design doc: https://gajanrapyutarobotics-my.sharepoint.com/:w:/g/personal/veeraj_khokale_rapyuta-robotics_com/EfxYaETn7ItNr5jGGRg4zjIBnXgWhDRmn5XFEhSS6Zr4EA?e=vkbIfr

- [x] Add unit test
- [x] Alica tests pass
- [x] Supplementary tests pass
- [x] Check turtlesim
- [x] Add support for default values
- [x] Add unit test for default values by regenerating etc folders using the new frontend

### CODE REVIEW

**Comment by veerajsk** [comment]
- [ ] Add unit test
- [ ] Existing tests work
- [ ] Check turtlesim

**Review by ash-anand — APPROVED** [review]

File: alica_engine/include/engine/AlicaContext.h [hunk 1]
```diff
@@ -302,8 +303,10 @@ class AlicaContext
      *
      * @return A reference to the global blackboard
      */
-    const Blackboard& getGlobalBlackboard() const;
-    Blackboard& editGlobalBlackboard();
+    [[deprecated("call std::shared_ptr<Blackboard> getGlobalBlackboard() instead")]] const Blackboard& getGlobalBlackboard() const;
```
Comment (bjoernschroeder) [inline alica_engine/include/engine/AlicaContext.h]: getGlobalBlackboardShared()

File: alica_engine/src/engine/AlicaContext.cpp [hunk 2]
```diff
@@ -25,11 +30,36 @@ AlicaContext::AlicaContext(const AlicaContextParams& alicaContextParams)
...
+        AlicaEngine::abort(LOGNAME, "Master file not found, aborting");
+    }
+    YAML::Node masterPlanNode;
+    try {
+        masterPlanNode = YAML::LoadFile(filePath);
+    } catch (YAML::BadFile& badFile) {
+        AlicaEngine::abort(LOGNAME, "Could not load master file: ", filePath, ", error details- ", badFile.msg, ", aborting");
+    }
+    auto inheritBlackboard = Factory::getValue<bool>(masterPlanNode, alica::Strings::inheritBlackboard, false);
+    if (inheritBlackboard && Factory::isValid(masterPlanNode[alica::Strings::blackboard])) {
```
Comment (bjoernschroeder) [inline alica_engine/src/engine/AlicaContext.cpp:57]: so the inheritBlackboard flag indicates that the master plan blackboard is the global blackboard? and if its not set, the maaster plan and the global blackboard will not be connected?

File: alica_tests/src/behaviours/TestInheritBlackboardBehaviour.cpp [hunk 3]
```diff
@@ -14,15 +14,14 @@ TestInheritBlackboardBehaviour::~TestInheritBlackboardBehaviour() {}
...
-    bb.set<int64_t>("masterKey", 3);
-    wm->passedParameters["masterKeyInBehavior"] = bb.get<int64_t>("masterKey");
-    if (bb.hasValue("behaviorKey")) {
-        wm->passedParameters["hasBehaviorKey"] = 3;
-    } else {
-        wm->passedParameters["hasBehaviorKey"] = 4;
+    int64_t masterKeyInBehavior = 0;
+    {
+        LockedBlackboardRW bb(*getBlackboard());
+        masterKeyInBehavior = bb.get<int64_t>("masterKey");
```
Comment (bjoernschroeder) [inline alica_tests/src/behaviours/TestInheritBlackboardBehaviour.cpp:20]: nit: std::int64_t

**Review by bjoernschroeder — COMMENTED** [review]
partial review, will continue tomorrow

File: alica_engine/src/engine/AlicaContext.cpp (hunk 2 above)
Comment (veerajsk) [reply alica_engine/src/engine/AlicaContext.cpp:57]: yes

**Comment by bjoernschroeder** [comment]
PR is looking good to me so far. It would be great to have tests with different types defined in the global blackboard and using default values when the type is std::any.

**Comment by veerajsk** [comment]
> PR is looking good to me so far. It would be great to have tests with different types defined in the global blackboard and using default values when the type is std::any.

we can't have default values for std::any because we won't be able to parse it without knowing the type

Now that the frontend & backend work, I have setup a default key in testGlobalBlackbordMaster.pml & the unit test basically tests that global blackboard work as expected. I don't think we need a test for each type because the code is essentially the same as for value mapping & we already have many tests for that.

**Review by bjoernschroeder — APPROVED** [review]

**Pull request merged**